### PR TITLE
Get timestamps during decoding

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR supports to get timestamps of each word during decoding. The files `recogs-*` that save decoding results would have lines to show the timestamp info of each decoded word in hyp.

E.g., 
```
1089-134686-0001-460-0: ref=['STUFF', 'IT', 'INTO', 'YOU', 'HIS', 'BELLY', 'COUNSELLED', 'HIM']                                                               
1089-134686-0001-460-0: hyp=['STUFF', 'IT', 'INTO', 'YOU', 'HIS', 'BELLY', 'COUNSELLED', 'HIM']                                                               
1089-134686-0001-460-0: timestamp=[0.6, 1.16, 1.44, 1.68, 2.04, 2.24, 2.64, 3.16]
```

### CODE REVIEW

**Comment by yaozengwei** [comment]
Refer to https://github.com/k2-fsa/sherpa/pull/119 in sherpa framework.

**Comment by yaozengwei** [comment]
Commit https://github.com/k2-fsa/icefall/pull/598/commits/6494e0fb790885e05f303e6fa8683ccd1be97229 supports to compute mean symbol delay with word-level alignments on the correctly predicted words. 

To compute symbol delay, you should set argument `--manifest-dir data/fbank_ali` which contains alignments to `decode.py`.  You will get results of computed symbol delay in log files and summary files `symbol-delay-summary-test-*`, as we have done for the WER. 

Screen Shot 2022-10-06 at 18 13 17

Note: If you set argument `--manifest-dir data/fbank` without alignments to `decode.py`, the shown symbol delay would be `inf`.

**Comment by yaozengwei** [comment]
The following table shows results of applying delay penalty https://github.com/k2-fsa/icefall/pull/573 on the lstm model
* trained on train-clean-100, decode with greedy_search

| epoch-40-avg-20 | WER | mean of symbol delay | variance of symbol delay | 
| - | - | - | - |
| baseline | 9.4 & 25.12 | 0.337s & 0.334s | 0.007 & 0.007 |
| with delay penalty = 0.006 | 10.11 & 26.38 | 0.217s & 0.227s | 0.006 & 0.007 |
| with delay penalty = 0.01 | 10.75 & 27.25 | 0.175s & 0.19s | 0.006 & 0.007 |

File: egs/librispeech/ASR/add_alignments.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
```
Comment (csukuangfj) [inline egs/librispeech/ASR/add_alignments.sh:2]: Could you add some doc describing what this file does?

File: egs/librispeech/ASR/local/add_alignment_librispeech.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import zipfile
+from pathlib import Path
+from typing import List
+
+from lhotse import CutSet, load_manifest_lazy
+from lhotse.recipes.librispeech import parse_alignments
+from lhotse.utils import is_module_available
+
+LIBRISPEECH_ALIGNMENTS_URL = (
+    "https://drive.google.com/uc?id=1WYfgr31T-PPwMcxuAq09XZfHQO5Mw8fE"
```
Comment (csukuangfj) [inline egs/librispeech/ASR/local/add_alignment_librispeech.py:36]: Some users from China may not have access to Google.
Could you make a copy of this data and upload it to huggingface?
You can leave a comment here saying that the alignment can also be found on huggingface.

File: egs/librispeech/ASR/local/add_alignment_librispeech.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        # add alignment attribute and write out
+        cuts_in = load_manifest_lazy(cuts_in_path)
+        with CutSet.open_writer(cuts_out_path) as writer:
+            for cut in cuts_in:
+                for idx, subcut in enumerate(cut.supervisions):
+                    origin_id = subcut.id.split("_")[0]
+                    if origin_id in alignments:
+                        ali = alignments[origin_id]
+                    else:
+                        ali = []
```
Comment (csukuangfj) [inline egs/librispeech/ASR/local/add_alignment_librispeech.py:168]: Could you print a warning for those cuts that don't have an alignment?

File: egs/librispeech/ASR/local/add_alignment_librispeech.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                    subcut.alignment = {"word": ali}
+                writer.write(cut, flush=True)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/local/add_alignment_librispeech.py:170]: I suggest that at the end we print the following information:
- IDs of cuts that don't have alignments
- IDs of alignments that don't have a corresponding cut.

File: egs/librispeech/ASR/lstm_transducer_stateless3/decode.py [hunk 5]
```diff
@@ -314,7 +316,7 @@ def decode_one_batch(
     batch: dict,
     word_table: Optional[k2.SymbolTable] = None,
     decoding_graph: Optional[k2.Fsa] = None,
-) -> Dict[str, List[List[str]]]:
+) -> Dict[str, Tuple[List[List[str]], List[List[float]]]]:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lstm_transducer_stateless3/decode.py:319]: please also update the docstring of this function.

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 6]
```diff
@@ -36,7 +42,8 @@ def fast_beam_search_one_best(
     max_states: int,
     max_contexts: int,
     temperature: float = 1.0,
-) -> List[List[int]]:
+    return_timestamps: bool = False,
+) -> Union[List[List[int]], DecodingResults]:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:46]: Please also update the docstring of the return value.

Comment (yaozengwei) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:46]: Thanks. I will update that.

**Comment by desh2608** [comment]
Do these time-stamps represent the start of the word? If yes, how do you obtain the duration (simply subtracting start of next work minus start of current word may not account for pauses)?

**Comment by yaozengwei** [comment]
The timestamp is the start of the each word. We can't get the end timestamp for each word in RNN-T model.

**Comment by csukuangfj** [comment]
> Do these time-stamps represent the start of the word?

Yes.

>  If yes, how do you obtain the duration (simply subtracting start of next word minus start of current word may not account for pauses

Yes.

For the transducer decoding, we can only get the start time of a word.

**Pull request merged**